marlowe-actus/src/Language/Marlowe/ACTUS/QCGenerator.hs

```diff
@@ -0,0 +1,180 @@
+{-# LANGUAGE DataKinds           #-}
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# OPTIONS_GHC -fno-warn-incomplete-uni-patterns -fno-warn-name-shadowing -fno-warn-unused-do-bind #-}
+module Language.Marlowe.ACTUS.QCGenerator where
+
+import           Language.Marlowe.ACTUS.Definitions.ContractTerms
+import           Test.QuickCheck
+import           Data.Time
+import           Data.Time.Clock.POSIX(posixSecondsToUTCTime)
+
+epochToDay :: Integer -> Day
+epochToDay = utctDay . posixSecondsToUTCTime . fromIntegral
+
+largeamount :: Gen Double
+largeamount = choose (0.0, 10000000.0)
+
+smallamount :: Gen Double
+smallamount = choose (0.0, 1000.0)
+
+percentage :: Gen Double
+percentage = choose (0.0, 100.0)
+
+scalingFactor :: Gen Double
+scalingFactor = choose (0.0, 100.0)
+
+mightbe :: Gen a -> Gen (Maybe a)
+mightbe original = oneof [ Just <$> original, return Nothing ]
+
+zeroOr :: Gen Double -> Gen Double
+zeroOr original = oneof [return 0.0, original]
+
+oneOr :: Gen Double -> Gen Double
+oneOr original = oneof [return 1.0, original]
+
+date :: Gen Day
+date = epochToDay <$> choose (0, 1000)
+
+datecycle :: Gen Cycle
+datecycle = Cycle <$> choose (0, 100) <*> elements [P_D, P_W, P_M, P_Q, P_H, P_Y] <*> elements [ShortStub, LongStub]
```

Comment:
It would be possible to somewhat correlate the size of the resulting contract to the value [sized](http://hackage.haskell.org/package/QuickCheck-2.14.2/docs/Test-QuickCheck-Gen.html#v:sized) from QuickCheck, that way the generation algorithm can adjust the size of the test cases, apparently QuickCheck starts trying small tests and then it adjusts the size overtime, so if something fails very often we will get a smaller counter-example (even without shrinking). I guess it would translate into making the `datecycle` larger, and the difference between start and end dates shorter, so that there are less cycles.

Reply:
I've rearranged that - although I don't take enddate - startdate into account, rather max(randomdate) - min(randomdate) for now.